File: apps/demo-be-python-fastapi/tests/integration/steps/reporting_steps.py
"""BDD step definitions for financial reporting feature."""

import json

from pytest_bdd import given, parsers, scenarios, then, when

from tests.integration.conftest import GHERKIN_ROOT
from tests.integration.service_client import FakeResponse, ServiceClient

scenarios(str(GHERKIN_ROOT / "expenses" / "reporting.feature"))

_PASSWORD = "Str0ng#Pass1"


@given(
    '"alice" has logged in and stored the access token',
    target_fixture="alice_tokens",
)
def alice_login_reporting(client: ServiceClient, registered_user: dict) -> dict:
    return client.login_user("alice", _PASSWORD)


@given(
    parsers.parse("alice has created an entry with body {body}"),
    target_fixture="reporting_entry",
)
def alice_create_reporting_entry(client: ServiceClient, alice_tokens: dict, body: str) -> dict:
    data = json.loads(body)
    resp = client.post_expense(f"Bearer {alice_tokens['access_token']}", data)
    assert resp.status_code == 201, f"Create entry failed: {resp.text}"
    return resp.json()


# --- When steps ---


@when(
    parsers.parse("alice sends GET /api/v1/reports/pl?from={from_}&to={to}&currency={currency}"),
    target_fixture="response",
)
def alice_get_pl_report(
    client: ServiceClient, alice_tokens: dict, from_: str, to: str, currency: str
) -> FakeResponse:
    return client.get_pl_report(
        f"Bearer {alice_tokens['access_token']}",
        from_,
        to,
        currency,
    )


# --- Then steps ---


@then(parsers.parse('the response body should contain "income_total" equal to "{value}"'))
def check_income_total(response: FakeResponse, value: str) -> None:
    body = response.json()
    assert "income_total" in body, f"income_total not in: {body}"
    assert str(body["income_total"]) == value, (
        f"Expected income_total={value}, got {body['income_total']}"
    )


@then(parsers.parse('the response body should contain "expense_total" equal to "{value}"'))
def check_expense_total(response: FakeResponse, value: str) -> None:
    body = response.json()
    assert "expense_total" in body, f"expense_total not in: {body}"
    assert str(body["expense_total"]) == value, (
        f"Expected expense_total={value}, got {body['expense_total']}"
    )


@then(parsers.parse('the response body should contain "net" equal to "{value}"'))
def check_net(response: FakeResponse, value: str) -> None:
    body = response.json()
    assert "net" in body, f"net not in: {body}"
    assert str(body["net"]) == value, f"Expected net={value}, got {body['net']}"


@then(parsers.parse('the income breakdown should contain "{category}" with amount "{amount}"'))
def check_income_breakdown(response: FakeResponse, category: str, amount: str) -> None:
    body = response.json()
    breakdown = body.get("income_breakdown", {})
    assert category in breakdown, f"Category '{category}' not in income_breakdown: {breakdown}"
    assert str(breakdown[category]) == amount, (
        f"Expected income_breakdown[{category}]={amount}, got {breakdown[category]}"
    )


@then(parsers.parse('the expense breakdown should contain "{category}" with amount "{amount}"'))
def check_expense_breakdown(response: FakeResponse, category: str, amount: str) -> None:
    body = response.json()
    breakdown = body.get("expense_breakdown", {})
    assert category in breakdown, f"Category '{category}' not in expense_breakdown: {breakdown}"
    assert str(breakdown[category]) == amount, (
        f"Expected expense_breakdown[{category}]={amount}, got {breakdown[category]}"
    )

File: apps/a-demo-be-elixir-phoenix/test/integration/steps/reporting_steps.exs
defmodule AADemoBeExphWeb.Integration.ReportingSteps do
  use Cabbage.Feature, async: false, file: "expenses/reporting.feature"

  use ADemoBeExph.DataCaseIntegration

  alias ADemoBeExph.Integration.Helpers
  alias ADemoBeExph.Integration.ServiceLayer

  @moduletag :integration

  defgiven ~r/^the API is running$/, _vars, state do
    {:ok, state}
  end

  defgiven ~r/^a user "(?<username>[^"]+)" is registered with email "(?<email>[^"]+)" and password "(?<password>[^"]+)"$/,
           %{username: username, email: email, password: password},
           state do
    user = Helpers.register_user!(username, email, password)
    {:ok, Map.put(state, :alice, user)}
  end

  defgiven ~r/^"alice" has logged in and stored the access token$/,
           _vars,
           %{alice: user} = state do
    {access_token, _} = Helpers.login_user!(user)
    {:ok, Map.put(state, :access_token, access_token)}
  end

  defgiven ~r/^alice has created an entry with body \{ (?<body>.+) \}$/,
           %{body: body_content},
           %{access_token: access_token} = state do
    params = Jason.decode!("{" <> body_content <> "}")
    response = ServiceLayer.create_expense(access_token, params)
    assert response.status == 201
    {:ok, state}
  end

  defwhen ~r/^alice sends GET \/api\/v1\/reports\/pl\?from=(?<from>[^&]+)&to=(?<to>[^&]+)&currency=(?<currency>[^\s]+)$/,
          %{from: from, to: to, currency: currency},
          %{access_token: access_token} = state do
    response = ServiceLayer.pl_report(access_token, from, to, currency)
    {:ok, Map.put(state, :response, response)}
  end

  defthen ~r/^the response status code should be (?<code>\d+)$/,
          %{code: code},
          %{response: response} = state do
    assert response.status == String.to_integer(code)
    {:ok, state}
  end

  defthen ~r/^the response body should contain "(?<field>[^"]+)" equal to "(?<value>[^"]+)"$/,
          %{field: field, value: value},
          %{response: response} = state do
    stored = response.body[field] |> Decimal.new() |> Decimal.to_string()
    expected = value |> Decimal.new() |> Decimal.to_string()
    assert stored == expected
    {:ok, state}
  end

  defthen ~r/^the income breakdown should contain "(?<category>[^"]+)" with amount "(?<amount>[^"]+)"$/,
          %{category: category, amount: amount},
          %{response: response} = state do
    income_breakdown = response.body["incomeBreakdown"]
    entry = Enum.find(income_breakdown, fn item -> item["category"] == category end)

    assert entry != nil,
           "Category '#{category}' not found in incomeBreakdown: #{inspect(income_breakdown)}"

    stored = entry["total"] |> Decimal.new() |> Decimal.to_string()
    expected = amount |> Decimal.new() |> Decimal.to_string()
    assert stored == expected
    {:ok, state}
  end

  defthen ~r/^the expense breakdown should contain "(?<category>[^"]+)" with amount "(?<amount>[^"]+)"$/,
          %{category: category, amount: amount},
          %{response: response} = state do
    expense_breakdown = response.body["expenseBreakdown"]
    entry = Enum.find(expense_breakdown, fn item -> item["category"] == category end)

    assert entry != nil,
           "Category '#{category}' not found in expenseBreakdown: #{inspect(expense_breakdown)}"

    stored = entry["total"] |> Decimal.new() |> Decimal.to_string()
    expected = amount |> Decimal.new() |> Decimal.to_string()
    assert stored == expected
    {:ok, state}
  end
end
